File: app/src/main/java/info/nightscout/androidaps/plugins/pump/insight/activities/InsightPairingActivity.java
package info.nightscout.androidaps.plugins.pump.insight.activities;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import info.nightscout.androidaps.R;
import info.nightscout.androidaps.activities.NoSplashAppCompatActivity;
import info.nightscout.androidaps.plugins.pump.insight.connection_service.InsightConnectionService;
import info.nightscout.androidaps.plugins.pump.insight.descriptors.InsightState;
import info.nightscout.androidaps.plugins.pump.insight.utils.ExceptionTranslator;

public class InsightPairingActivity extends NoSplashAppCompatActivity implements InsightConnectionService.StateCallback, View.OnClickListener, InsightConnectionService.ExceptionCallback {

    private boolean scanning;
    private LinearLayout deviceSearchSection;
    private TextView pleaseWaitSection;
    private LinearLayout codeCompareSection;
    private LinearLayout pairingCompletedSection;
    private Button yes;
    private Button no;
    private TextView code;
    private Button exit;
    private RecyclerView deviceList;
    private DeviceAdapter deviceAdapter = new DeviceAdapter();

    private InsightConnectionService service;

    private ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder binder) {
            service = ((InsightConnectionService.LocalBinder) binder).getService();
            if (service.isPaired()) return;
            else {
                service.requestConnection(InsightPairingActivity.this);
                service.registerStateCallback(InsightPairingActivity.this);
                service.registerExceptionCallback(InsightPairingActivity.this);
                onStateChanged(service.getState());
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {

        }
    };

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_insight_pairing);

        deviceSearchSection = findViewById(R.id.device_search_section);
        pleaseWaitSection = findViewById(R.id.please_wait_section);
        codeCompareSection = findViewById(R.id.code_compare_section);
        pairingCompletedSection = findViewById(R.id.pairing_completed_section);
        yes = findViewById(R.id.yes);
        no = findViewById(R.id.no);
        code = findViewById(R.id.code);
        exit = findViewById(R.id.exit);
        deviceList = findViewById(R.id.device_list);

        yes.setOnClickListener(this);
        no.setOnClickListener(this);
        exit.setOnClickListener(this);

        deviceList.setLayoutManager(new LinearLayoutManager(this));
        deviceList.setAdapter(deviceAdapter);


        bindService(new Intent(this, InsightConnectionService.class), serviceConnection, BIND_AUTO_CREATE);
}

    @Override
    protected void onDestroy() {
        if (service != null) {
            service.withdrawConnectionRequest(InsightPairingActivity.this);
            service.unregisterStateCallback(InsightPairingActivity.this);
            service.unregisterExceptionCallback(InsightPairingActivity.this);
        }
        unbindService(serviceConnection);
        super.onDestroy();
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (service != null && service.getState() == InsightState.NOT_PAIRED) startBLScan();
    }

    @Override
    protected void onStop() {
        stopBLScan();
        super.onStop();
    }

    @Override
    public void onStateChanged(InsightState state) {
        runOnUiThread(() -> {
            switch (state) {
                case NOT_PAIRED:
                    startBLScan();
                    deviceSearchSection.setVisibility(View.VISIBLE);
                    pleaseWaitSection.setVisibility(View.GONE);
                    codeCompareSection.setVisibility(View.GONE);
                    pairingCompletedSection.setVisibility(View.GONE);
                    break;
                case CONNECTING:
                case SATL_CONNECTION_REQUEST:
                case SATL_KEY_REQUEST:
                case SATL_VERIFY_DISPLAY_REQUEST:
                case SATL_VERIFY_CONFIRM_REQUEST:
                case APP_BIND_MESSAGE:
                    stopBLScan();
                    deviceSearchSection.setVisibility(View.GONE);
                    pleaseWaitSection.setVisibility(View.VISIBLE);
                    codeCompareSection.setVisibility(View.GONE);
                    pairingCompletedSection.setVisibility(View.GONE);
                    break;
                case AWAITING_CODE_CONFIRMATION:
                    stopBLScan();
                    deviceSearchSection.setVisibility(View.GONE);
                    pleaseWaitSection.setVisibility(View.GONE);
                    codeCompareSection.setVisibility(View.VISIBLE);
                    pairingCompletedSection.setVisibility(View.GONE);
                    code.setText(service.getVerificationString());
                    break;
                case DISCONNECTED:
                case CONNECTED:
                    stopBLScan();
                    deviceSearchSection.setVisibility(View.GONE);
                    pleaseWaitSection.setVisibility(View.GONE);
                    codeCompareSection.setVisibility(View.GONE);
                    pairingCompletedSection.setVisibility(View.VISIBLE);
                    break;
            }
        });
    }

    private void startBLScan() {
        if (!scanning) {
            BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
            if (bluetoothAdapter != null) {
                if (!bluetoothAdapter.isEnabled()) bluetoothAdapter.enable();
                IntentFilter intentFilter = new IntentFilter();
                intentFilter.addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED);
                intentFilter.addAction(BluetoothDevice.ACTION_FOUND);
                registerReceiver(broadcastReceiver, intentFilter);
                bluetoothAdapter.startDiscovery();
                scanning = true;
            }
        }
    }

    private void stopBLScan() {
        if (scanning) {
            unregisterReceiver(broadcastReceiver);
            BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
            if (bluetoothAdapter != null) {
                bluetoothAdapter.cancelDiscovery();
            }
            scanning = false;
        }
    }
    @Override
    public void onClick(View v) {
        if (v == exit) finish();
        else if (v == yes) service.confirmVerificationString();
        else if (v == no) service.rejectVerificationString();
    }

    @Override
    public void onExceptionOccur(Exception e) {
        ExceptionTranslator.makeToast(this, e);
    }

    private void deviceSelected(BluetoothDevice device) {
        service.pair(device.getAddress());
    }

    private BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String action = intent.getAction();
            if (action.equals(BluetoothAdapter.ACTION_DISCOVERY_FINISHED))
                BluetoothAdapter.getDefaultAdapter().startDiscovery();
            else if (action.equals(BluetoothDevice.ACTION_FOUND)) {
                BluetoothDevice bluetoothDevice = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);
                deviceAdapter.addDevice(bluetoothDevice);
            }
        }
    };

    private class DeviceAdapter extends RecyclerView.Adapter<DeviceAdapter.ViewHolder> {

        private List<BluetoothDevice> bluetoothDevices = new ArrayList<>();

        public void addDevice(BluetoothDevice bluetoothDevice) {
            if (!bluetoothDevices.contains(bluetoothDevice)) {
                bluetoothDevices.add(bluetoothDevice);
                notifyDataSetChanged();
            }
        }

        public void clear() {
            bluetoothDevices.clear();
            notifyDataSetChanged();
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.bluetooth_device, parent, false));
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            BluetoothDevice bluetoothDevice = bluetoothDevices.get(position);
            holder.deviceName.setText(bluetoothDevice.getName() == null ? bluetoothDevice.getAddress() : bluetoothDevice.getName());
            holder.deviceName.setOnClickListener((v) -> deviceSelected(bluetoothDevice));
        }

        @Override
        public int getItemCount() {
            return bluetoothDevices.size();
        }

        public class ViewHolder extends RecyclerView.ViewHolder {

            private TextView deviceName;

            public ViewHolder(View itemView) {
                super(itemView);
                deviceName = (TextView) itemView;
            }
        }

    }
}

File: insight/src/main/java/info/nightscout/androidaps/plugins/pump/insight/activities/InsightPairingActivity.kt
package info.nightscout.androidaps.plugins.pump.insight.activities

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.content.*
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import info.nightscout.androidaps.activities.NoSplashAppCompatActivity
import info.nightscout.androidaps.insight.R
import info.nightscout.androidaps.insight.databinding.ActivityInsightPairingBinding
import info.nightscout.androidaps.interfaces.PumpSync
import info.nightscout.androidaps.plugins.pump.common.ble.BlePreCheck
import info.nightscout.androidaps.plugins.pump.insight.connection_service.InsightConnectionService
import info.nightscout.androidaps.plugins.pump.insight.connection_service.InsightConnectionService.ExceptionCallback
import info.nightscout.androidaps.plugins.pump.insight.descriptors.InsightState
import info.nightscout.androidaps.plugins.pump.insight.utils.ExceptionTranslator
import javax.inject.Inject

class InsightPairingActivity : NoSplashAppCompatActivity(), InsightConnectionService.StateCallback, View.OnClickListener, ExceptionCallback {

    @Inject lateinit var blePreCheck: BlePreCheck
    @Inject lateinit var context: Context
    @Inject lateinit var pumpSync: PumpSync

    private lateinit var binding: ActivityInsightPairingBinding
    private var scanning = false
    private val deviceAdapter = DeviceAdapter()

    private val PERMISSION_REQUEST_BLUETOOTH = 30242

    private var service: InsightConnectionService? = null
    private val serviceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, binder: IBinder) {
            service = (binder as InsightConnectionService.LocalBinder).service
            service?.let {
                if (it.isPaired)
                    return
                else {
                    it.requestConnection(this@InsightPairingActivity)
                    it.registerStateCallback(this@InsightPairingActivity)
                    it.registerExceptionCallback(this@InsightPairingActivity)
                    onStateChanged(it.state)
                    pumpSync.connectNewPump()
                }
            }
        }

        override fun onServiceDisconnected(name: ComponentName) {}
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityInsightPairingBinding.inflate(layoutInflater)
        setContentView(binding.root)
        blePreCheck.prerequisitesCheck(this)
        binding.yes.setOnClickListener(this)
        binding.no.setOnClickListener(this)
        binding.exit.setOnClickListener(this)
        binding.deviceList.setLayoutManager(LinearLayoutManager(this))
        binding.deviceList.setAdapter(deviceAdapter)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (ContextCompat.checkSelfPermission(context, "android.permission.BLUETOOTH_CONNECT") != PackageManager.PERMISSION_GRANTED ||
                ContextCompat.checkSelfPermission(context, "android.permission.BLUETOOTH_SCAN") != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(this, arrayOf("android.permission.BLUETOOTH_SCAN", "android.permission.BLUETOOTH_CONNECT"), PERMISSION_REQUEST_BLUETOOTH)
                finish()
                return
            }
        }
        bindService(Intent(this, InsightConnectionService::class.java), serviceConnection, BIND_AUTO_CREATE)
    }

    override fun onDestroy() {
        service?.run {
            withdrawConnectionRequest(this@InsightPairingActivity)
            unregisterStateCallback(this@InsightPairingActivity)
            unregisterExceptionCallback(this@InsightPairingActivity)
        }
        unbindService(serviceConnection)
        super.onDestroy()
    }

    override fun onStart() {
        super.onStart()
        if (service != null && service!!.state === InsightState.NOT_PAIRED) startBLScan()
    }

    override fun onStop() {
        stopBLScan()
        super.onStop()
    }

    override fun onStateChanged(state: InsightState?) {
        runOnUiThread {
            when (state) {
                InsightState.NOT_PAIRED                             -> {
                    startBLScan()
                    binding.deviceSearchSection.visibility = View.VISIBLE
                    binding.pleaseWaitSection.visibility = View.GONE
                    binding.codeCompareSection.visibility = View.GONE
                    binding.pairingCompletedSection.visibility = View.GONE
                }

                InsightState.CONNECTING,
                InsightState.SATL_CONNECTION_REQUEST,
                InsightState.SATL_KEY_REQUEST,
                InsightState.SATL_VERIFY_DISPLAY_REQUEST,
                InsightState.SATL_VERIFY_CONFIRM_REQUEST,
                InsightState.APP_BIND_MESSAGE                       -> {
                    stopBLScan()
                    binding.deviceSearchSection.visibility = View.GONE
                    binding.pleaseWaitSection.visibility = View.VISIBLE
                    binding.codeCompareSection.visibility = View.GONE
                    binding.pairingCompletedSection.visibility = View.GONE
                }

                InsightState.AWAITING_CODE_CONFIRMATION             -> {
                    stopBLScan()
                    binding.deviceSearchSection.visibility = View.GONE
                    binding.pleaseWaitSection.visibility = View.GONE
                    binding.codeCompareSection.visibility = View.VISIBLE
                    binding.pairingCompletedSection.visibility = View.GONE
                    binding.code.text = service!!.verificationString
                }

                InsightState.DISCONNECTED, InsightState.CONNECTED   -> {
                    stopBLScan()
                    binding.deviceSearchSection.visibility = View.GONE
                    binding.pleaseWaitSection.visibility = View.GONE
                    binding.codeCompareSection.visibility = View.GONE
                    binding.pairingCompletedSection.visibility = View.VISIBLE
                }

                else                                                -> Unit
            }
        }
    }

    private fun startBLScan() {
        if (!scanning) {
            val bluetoothAdapter = (context.getSystemService(BLUETOOTH_SERVICE) as BluetoothManager).adapter
            if (bluetoothAdapter != null) {
                if (!bluetoothAdapter.isEnabled) bluetoothAdapter.enable()
                val intentFilter = IntentFilter()
                intentFilter.addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED)
                intentFilter.addAction(BluetoothDevice.ACTION_FOUND)
                registerReceiver(broadcastReceiver, intentFilter)
                bluetoothAdapter.startDiscovery()
                scanning = true
            }
        }
    }

    private fun stopBLScan() {
        if (scanning) {
            unregisterReceiver(broadcastReceiver)
            val bluetoothAdapter = (context.getSystemService(BLUETOOTH_SERVICE) as BluetoothManager).adapter
            bluetoothAdapter?.cancelDiscovery()
            scanning = false
        }
    }

    override fun onClick(v: View) {
        if (v === binding.exit) finish() else if (v === binding.yes) service!!.confirmVerificationString() else if (v === binding.no) service!!.rejectVerificationString()
    }

    override fun onExceptionOccur(e: Exception?) {
        e?.let { ExceptionTranslator.makeToast(this, it) }
    }

    private fun deviceSelected(device: BluetoothDevice) {
        service?.pair(device.address)
    }

    private val broadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            if (action == BluetoothAdapter.ACTION_DISCOVERY_FINISHED)
                (context.getSystemService(BLUETOOTH_SERVICE) as BluetoothManager).adapter.startDiscovery()
            else if (action == BluetoothDevice.ACTION_FOUND) {
                val bluetoothDevice = intent.getParcelableExtra<BluetoothDevice>(BluetoothDevice.EXTRA_DEVICE)
                bluetoothDevice?.let { deviceAdapter.addDevice(it) }
            }
        }
    }

    private inner class DeviceAdapter : RecyclerView.Adapter<DeviceAdapter.ViewHolder>() {

        private val bluetoothDevices: MutableList<BluetoothDevice> = ArrayList()
        fun addDevice(bluetoothDevice: BluetoothDevice) {
            if (!bluetoothDevices.contains(bluetoothDevice)) {
                bluetoothDevices.add(bluetoothDevice)
                notifyDataSetChanged()
            }
        }

        fun clear() {
            bluetoothDevices.clear()
            notifyDataSetChanged()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.bluetooth_device, parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val bluetoothDevice = bluetoothDevices[position]
            holder.deviceName.text = if (bluetoothDevice.name == null) bluetoothDevice.address else bluetoothDevice.name
            holder.deviceName.setOnClickListener { deviceSelected(bluetoothDevice) }
        }

        override fun getItemCount(): Int {
            return bluetoothDevices.size
        }

        inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

            val deviceName: TextView

            init {
                deviceName = itemView as TextView
            }
        }
    }
}